File: Prototype 2/danger.py
import os
import shutil
import sys
import uuid
import logging
from tqdm import tqdm
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.backends import default_backend


# ---------------- Logger Setup ---------------- #
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - [%(levelname)s] %(message)s",
    handlers=[logging.StreamHandler(sys.stdout)]
)
logger = logging.getLogger("DiskWiper")


# ---------------- Encryption Utility ---------------- #
def encrypt_chunk(data, key, iv):
    """Encrypt a block of data with AES-256-CBC."""
    cipher = Cipher(algorithms.AES(key), modes.CBC(iv), backend=default_backend())
    encryptor = cipher.encryptor()
    # Pad to 16 bytes
    if len(data) % 16 != 0:
        data += b" " * (16 - len(data) % 16)
    return encryptor.update(data) + encryptor.finalize()


# ---------------- Disk Wiper ---------------- #
def wipe_free_space(drive_root, passes=3, chunk_size=512 * 1024 * 1024):
    """
    Securely wipe free disk space by filling it with patterns and encrypted data.
    
    Args:
        drive_root (str): Path to the drive root (e.g., 'C:\\' on Windows or '/' on Linux).
        passes (int): Number of overwrite passes.
        chunk_size (int): Write size in bytes (default: 512MB).
    """
    logger.info(f"Starting secure free space wipe on: {drive_root}")
    total_free = shutil.disk_usage(drive_root).free
    logger.info(f"Total free space detected: {total_free // (1024**3)} GB")

    patterns = [
        (b"\x00", "zeros"),
        (b"\xFF", "ones"),
        (None, "random"),
        ("encrypted", "AES-256 encrypted random data"),
    ]

    for p in range(passes):
        logger.info(f"=== Pass {p+1}/{passes} ===")
        file_index = 0
        temp_files = []

        for pattern, label in patterns:
            logger.info(f"Writing pass pattern: {label}")

            try:
                while True:
                    temp_path = os.path.join(drive_root, f"shred_temp_{p}_{file_index}.dat")
                    file_index += 1
                    temp_files.append(temp_path)
                    with open(temp_path, "wb") as f:
                        written = 0
                        pbar = tqdm(total=total_free, unit="B", unit_scale=True, desc=f"Filling {label}")

                        while written < total_free:
                            ...
                            written += len(chunk)
                            pbar.update(len(chunk))

                        pbar.close()
                        logger.info(f"Temp file {temp_path} size: {written // (1024**3)} GB ({written} bytes)")

                    with open(temp_path, "wb") as f:
                        written = 0
                        pbar = tqdm(total=total_free, unit="B", unit_scale=True, desc=f"Filling {label}")

                        while written < total_free:
                            if pattern == "encrypted":
                                key = os.urandom(32)
                                iv = os.urandom(16)
                                data = os.urandom(chunk_size)
                                chunk = encrypt_chunk(data, key, iv)
                            elif pattern is None:
                                chunk = os.urandom(chunk_size)
                            else:
                                chunk = pattern * chunk_size

                            try:
                                f.write(chunk)
                                f.flush()
                                os.fsync(f.fileno())
                            except OSError:
                                # Out of space
                                break

                            written += len(chunk)
                            pbar.update(len(chunk))

                        pbar.close()

            except OSError:
                logger.warning("Disk is full for this pattern.")
            finally:
                # Delete temporary files to free space for next pass
                logger.info("Cleaning up temporary files...")
                for tf in temp_files:
                    try:
                        if os.path.exists(tf):
                            os.remove(tf)
                    except Exception as e:
                        logger.error(f"Failed to remove {tf}: {e}")
                temp_files.clear()

    logger.info("✅ Free space wiping completed successfully.")


# ---------------- CLI Entry ---------------- #
if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python disk_wiper.py <drive_root> [passes]")
        print("Example (Windows): python disk_wiper.py C:\\ 3")
        print("Example (Linux):   sudo python3 disk_wiper.py / 3")
        sys.exit(1)

    drive_root = sys.argv[1]
    passes = int(sys.argv[2]) if len(sys.argv) > 2 else 3

    wipe_free_space(drive_root, passes=passes)

File: Prototype 2/danger.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/statvfs.h>
#include <errno.h>
#include <time.h>
#include <openssl/evp.h>

// ---------- Logging ----------
#define LOG(fmt, ...) fprintf(stdout, "[*] " fmt "\n", ##__VA_ARGS__)
#define ERR(fmt, ...) fprintf(stderr, "[!] " fmt " (errno=%d: %s)\n", ##__VA_ARGS__, errno, strerror(errno))

// ---------- Encryption ----------
int encrypt_chunk(unsigned char *data, size_t len, unsigned char *out, unsigned char *key, unsigned char *iv) {
    EVP_CIPHER_CTX *ctx = EVP_CIPHER_CTX_new();
    if (!ctx) return -1;

    int outlen1, outlen2;
    if (EVP_EncryptInit_ex(ctx, EVP_aes_256_cbc(), NULL, key, iv) != 1) return -1;

    if (EVP_EncryptUpdate(ctx, out, &outlen1, data, (int)len) != 1) return -1;
    if (EVP_EncryptFinal_ex(ctx, out + outlen1, &outlen2) != 1) return -1;

    EVP_CIPHER_CTX_free(ctx);
    return outlen1 + outlen2;
}

// ---------- Disk Wiper ----------
void wipe_free_space(const char *path, int passes, size_t chunk_size) {
    struct statvfs fs;
    if (statvfs(path, &fs) != 0) {
        ERR("statvfs failed");
        return;
    }

    unsigned long long free_space = fs.f_bsize * fs.f_bavail;
    LOG("Detected free space: %llu MB", free_space / (1024 * 1024));

    for (int p = 0; p < passes; p++) {
        LOG("=== Pass %d/%d ===", p + 1, passes);

        const char *patterns[] = {"zeros", "ones", "random", "encrypted"};
        for (int pat = 0; pat < 4; pat++) {
            char filename[256];
            snprintf(filename, sizeof(filename), "%s/shred_temp_%d_%d.dat", path, p, pat);

            LOG("Writing pattern: %s -> %s", patterns[pat], filename);

            int fd = open(filename, O_CREAT | O_WRONLY | O_TRUNC, 0600);
            if (fd < 0) {
                ERR("Failed to open temp file");
                continue;
            }

            unsigned char *buf = malloc(chunk_size);
            unsigned char *outbuf = malloc(chunk_size + 32);
            if (!buf || !outbuf) {
                ERR("Memory allocation failed");
                close(fd);
                free(buf);
                free(outbuf);
                continue;
            }

            size_t written = 0;
            while (1) {
                if (strcmp(patterns[pat], "zeros") == 0) {
                    memset(buf, 0x00, chunk_size);
                } else if (strcmp(patterns[pat], "ones") == 0) {
                    memset(buf, 0xFF, chunk_size);
                } else if (strcmp(patterns[pat], "random") == 0) {
                    for (size_t i = 0; i < chunk_size; i++) buf[i] = rand() % 256;
                } else if (strcmp(patterns[pat], "encrypted") == 0) {
                    for (size_t i = 0; i < chunk_size; i++) buf[i] = rand() % 256;
                    unsigned char key[32], iv[16];
                    for (int i = 0; i < 32; i++) key[i] = rand() % 256;
                    for (int i = 0; i < 16; i++) iv[i] = rand() % 256;
                    int enc_size = encrypt_chunk(buf, chunk_size, outbuf, key, iv);
                    if (enc_size > 0) memcpy(buf, outbuf, enc_size);
                }

                ssize_t w = write(fd, buf, chunk_size);
                if (w <= 0) {
                    break; // likely disk full
                }
                written += w;
            }

            LOG("Wrote %zu MB", written / (1024 * 1024));

            close(fd);
            free(buf);
            free(outbuf);

            // Delete temp file
            if (remove(filename) != 0) {
                ERR("Failed to remove temp file");
            } else {
                LOG("Removed %s", filename);
            }
        }
    }

    LOG("✅ Free space wipe complete.");
}

// ---------- Main ----------
int main(int argc, char *argv[]) {
    if (argc < 2) {
        printf("Usage: %s <path> [passes]\n", argv[0]);
        printf("Example (Linux):   sudo %s / 3\n", argv[0]);
        printf("Example (Windows): %s C:/ 3\n", argv[0]);
        return 1;
    }

    const char *path = argv[1];
    int passes = (argc > 2) ? atoi(argv[2]) : 3;
    size_t chunk_size = 512 * 1024 * 1024; // 512MB

    srand((unsigned int)time(NULL));
    wipe_free_space(path, passes, chunk_size);
    return 0;
}
